adafruit_led_animation/animation.py

```diff
@@ -0,0 +1,585 @@
+        self._tail_length = tail_length + 1
+        self._color_step = 0.9 / tail_length
+        self._color_offset = 0.1
+        self._comet_colors = None
+        self._reverse_comet_colors = None
+        self.reverse = reverse
+        self.bounce = bounce
+        super(Comet, self).__init__(pixel_object, speed, color)
+        self._generator = self._comet_generator()
+
+    def _recompute_color(self, color):
+        self._comet_colors = [BLACK] + [
+            [int(color[rgb] * ((n * self._color_step) + self._color_offset))
+             for rgb in range(len(color))
+            ] for n in range(self._tail_length - 1)
+        ]
+        self._reverse_comet_colors = list(reversed(self._comet_colors))
+
+    def _get_range(self, num_pixels):
+        if self.reverse:
+            return range(num_pixels, -self._tail_length - 1, -1)
+        return range(-self._tail_length, num_pixels + 1)
+
+    def _comet_generator(self):
+        num_pixels = len(self.pixel_object)
+        while True:
+            colors = self._reverse_comet_colors if self.reverse else self._comet_colors
+            for start in self._get_range(num_pixels):
+
+                if start + self._tail_length < num_pixels:
+                    end = self._tail_length
+                else:
+                    end = num_pixels - start
+                if start <= 0:
+                    num_visible = self._tail_length + start
+                    self.pixel_object[0:num_visible] = colors[self._tail_length - num_visible:]
+                else:
+                    self.pixel_object[start:start + end] = colors[0:end]
+                self.show()
+                yield
+            if self.bounce:
+                self.reverse = not self.reverse
+
+    def draw(self):
+        next(self._generator)
+
+
+class Sparkle(Animation):
+    """
+    Sparkle animation of a single color.
+
+    :param pixel_object: The initialised LED object.
+    :param int speed: Animation speed in seconds, e.g. ``0.1``.
+    :param color: Animation color in ``(r, g, b)`` tuple, or ``0x000000`` hex format.
+    """
+    def __init__(self, pixel_object, speed, color):
+        self._half_color = None
+        self._dim_color = None
+        super(Sparkle, self).__init__(pixel_object, speed, color)
+
+    def _recompute_color(self, color):
+        half_color = tuple(color[rgb] // 4 for rgb in range(len(color)))
+        dim_color = tuple(color[rgb] // 10 for rgb in range(len(color)))
+        for pixel in range(len(self.pixel_object)):
+            if self.pixel_object[pixel] == self._half_color:
+                self.pixel_object[pixel] = half_color
+            elif self.pixel_object[pixel] == self._dim_color:
+                self.pixel_object[pixel] = dim_color
+        self._half_color = half_color
+        self._dim_color = dim_color
+
+    def draw(self):
+        pixel = random.randint(0, (len(self.pixel_object) - 2))
```

Comment:
what if the pixel object has less than 2 pixels?

Reply:
Then it would crash.  Good call.  (note: this was based on code from elsewhere, and this animation doesn't support single pixels -- __init__ should check that).

Reply:
Init now checks that the strip is big enough.